hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/table/format/CastMap.java

```diff
@@ -165,21 +192,132 @@ void add(int pos, LogicalType fromType, LogicalType toType) {
+      Object toObject = fromObject != null ? getConversion(fromType, toType).apply(fromObject) : null;
+      objects[i] = toObject;
+    }
+    return new GenericArrayData(objects);
+  }
+
+  /**
+   * Helper function to perform convert a MapData from one LogicalType to another.
+   *
+   * @param map      Non-null map data to be converted; however, values are allowed to be null
+   * @param fromType The input LogicalType of the row data to be converted from
+   * @param toType   The output LogicalType of the row data to be converted to
+   * @return Converted map that has the structure/specifications of that defined by the output LogicalType
+   */
+  private static MapData doMapConversion(@Nonnull MapData map, LogicalType fromType, LogicalType toType) {
+    // no schema evolution is allowed on the keyType, hence, we only need to care about the valueType
+    LogicalType fromValueType = fromType.getChildren().get(1);
+    LogicalType toValueType = toType.getChildren().get(1);
+    LogicalType keyType = fromType.getChildren().get(0);
+
+    final Map<Object, Object> result = new HashMap<>();
+    for (int i = 0; i < map.size(); i++) {
+      Object keyObject = ArrayData.createElementGetter(keyType).getElementOrNull(map.keyArray(), i);
+      Object fromObject = ArrayData.createElementGetter(fromValueType).getElementOrNull(map.valueArray(), i);
+      // need to handle nulls to prevent NullPointerException in #getConversion()
+      Object toObject = fromObject != null ? getConversion(fromValueType, toValueType).apply(fromObject) : null;
+      result.put(keyObject, toObject);
+    }
+    return new GenericMapData(result);
+  }
+
+  /**
+   * Helper function to perform convert a RowData from one LogicalType to another.
+   *
+   * @param row      Non-null row data to be converted; however, fields might contain nulls
+   * @param fromType The input LogicalType of the row data to be converted from
+   * @param toType   The output LogicalType of the row data to be converted to
+   * @return Converted row that has the structure/specifications of that defined by the output LogicalType
+   */
+  private static RowData doRowConversion(@Nonnull RowData row, LogicalType fromType, LogicalType toType) {
+    // note: InternalSchema.merge guarantees that the schema to be read fromType is orientated in the same order as toType
+    // hence, we can match types by position as it is guaranteed that it is referencing the same field
+    List<LogicalType> fromChildren = fromType.getChildren();
+    List<LogicalType> toChildren = toType.getChildren();
+    ValidationUtils.checkArgument(fromChildren.size() == toChildren.size(),
+        "fromType [" + fromType + "] size: != toType [" + toType + "] size");
+
+    GenericRowData rowData = new GenericRowData(toType.getChildren().size());
+    for (int i = 0; i < toChildren.size(); i++) {
+      Object fromVal = RowData.createFieldGetter(fromChildren.get(i), i).getFieldOrNull(row);
+      Object toVal;
```

Comment:
caution for the performance, because you are constructing the field getter for each row->row conversion.

Reply:
Any suggestions on how we can work around this? Don't think this is avoidable as of now.

Reply:
Maybe add a wrapper class like what we to in `RowDataToAvroConverters` for row data and avro conversion.

Reply:
Done, please take a look. 

I also added a check in the `testCastNestedRow` case in `TestCastMap` to test  `INT -> INT NOT NULL` conversions as I noticed we are not handling such cases in the past.